source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+     */
+    __m256i maskA = _mm256_set1_epi32(0xFF); // low bits
+    __m256i maskB = _mm256_set1_epi32(0xFF00);
+    __m256i maskC = _mm256_set1_epi32(0xFF0000);
+    __m256i maskD = _mm256_set1_epi32(0xFF000000);
+
+    __m256i bitsA = _mm256_slli_epi32(in & maskA, 18);
+    __m256i bitsB = _mm256_slli_epi32(in & maskB, 4);
+    __m256i bitsC = _mm256_srli_epi32(in & maskC, 10);
+    __m256i bitsD = _mm256_srli_epi32(in & maskD, 24);
+
+    __m256i dwords = (bitsA | bitsB) | (bitsC | bitsD);
+    /*
+     * Now we have a series of dwords with empty MSBs.
+     * We need to pack them together (and shift down) with a shuffle operation.
+     * Unfortunately the shuffle operation operates independently within each 128-bit lane,
+     * so we'll need to do this in two steps: First we compact dwords within each lane, then
+     * we do a dword shuffle to compact the two lanes together.
+
+     * 15 14 13 12 11 10 09 08 07 06 05 04 03 02 01 00 <- byte index (little endian)
+     * -- 09 0a 0b -- 06 07 08 -- 03 04 05 -- 00 01 02 <- data index
+     *
+     * We also reverse the order of 3-byte fragments within each lane; we've constructed
+     * those fragments in little endian but the order of fragments within the overall
+     * vector is in memory order (big endian)
+     */
+    const struct aligned256 shufvec_buf = {{
+        /* clang-format off */
+        /* MSB */
+        0xFF, 0xFF, 0xFF, 0xFF, /* Zero out the top 4 bytes of the lane */
+        2,  1,  0,
+        6,  5,  4,
+        10,  9,  8,
+        14, 13, 12,
+
+        0xFF, 0xFF, 0xFF, 0xFF, /* Zero out the top 4 bytes of the lane */
+        2,  1,  0,
+        6,  5,  4,
+        10,  9,  8,
+        14, 13, 12
+        /* LSB */
+        /* clang-format on */
+    }};
+    __m256i shufvec = _mm256_load_si256((__m256i const *)&shufvec_buf);
+
+    dwords = _mm256_shuffle_epi8(dwords, shufvec);
+    /*
+     * Now shuffle the 32-bit words:
+     * A B C 0 D E F 0 -> 0 0 A B C D E F
+     */
+    __m256i shuf32 = _mm256_set_epi32(0, 0, 7, 6, 5, 3, 2, 1);
+
+    dwords = _mm256_permutevar8x32_epi32(dwords, shuf32);
+
+    return dwords;
+}
+
+static inline bool decode(const unsigned char *in, unsigned char *out) {
+    __m256i vec = _mm256_loadu_si256((__m256i const *)in);
+    if (!decode_vec(&vec)) {
+        return false;
+    }
+    vec = pack_vec(vec);
+
+    /*
+     * We'll do overlapping writes to get both the low 128 bits and the high 64-bits written.
+     * Input (memory order): 0 1 2 3 4 5 - - (dwords)
+     * Input (little endian) - - 5 4 3 2 1 0
+     * Output in memory:
+     * [0 1 2 3] [4 5]
+     */
+    __m128i lo = _mm256_extracti128_si256(vec, 0);
+    uint64_t hi = _mm256_extract_epi64(vec, 2);
+
+    _mm_storeu_si128((__m128i *)out, lo);
+    memcpy(out + 16, &hi, sizeof(hi));
+
+    return true;
+}
+
+size_t aws_common_private_base64_decode_sse41(const unsigned char *in, unsigned char *out, size_t len) {
+    if (len % 4)
```

Comment:
I prolly wrote this, but since we're on a perfomance kick here:

if (len & 0x03) {
    return AWS_OP_ERR;
}

Reply:
Any decent optimizing compiler will deal with this for us.

Reply:
I'm with Bryan on this one, `len % 4` is much clearer to the reader and per this [godbolt](https://godbolt.org/z/cDyOhw) instance, becomes a `test len, 3` at just -O1.